app/helpers/cache_helper.rb

```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+module CacheHelper
+  # Yields a cached query, expired by the most recently updated record of a given class
+  def cached_data_by_class(cache_key, cached_class)
+    Rails.cache.fetch "#{cached_class}-#{cache_key}-#{latest_timestamp_by_class(cached_class)}" do
+      yield
+    end
+  end
+
+  # Gets the updated_at timestamp of the most recently updated record of a given class
+  def latest_timestamp_by_class(cached_class)
+    @_memoized_timestamps ||= Hash.new do |hash, key|
+      hash[key] = key.maximum(:updated_at).to_i
+    end
+    @_memoized_timestamps[cached_class]
+  end
```

Comment:
I don't see what's the actual benefit of this approach compared to Rail's [compound cache key](https://guides.rubyonrails.org/caching_with_rails.html#fragment-caching). There the expiration happens by simply fetching a new cache key whereas we have to check all the models' `update_at`.

Reply:
Rails's cache key applies to a single record, right? Like `cache: @post`. Here we're creating a usable cache key for all the records for a class, like `Spree::Taxon`, and it expires when any taxon is updated. Is there a built-in Rails way to do that? I wasn't trying to avoid `touch` here.

I added memoization afterwards because there can be multiple calls to the method for the same class in a request cycle, so we can avoid re-fetching the same value for the same class. The initial method (without the memoization) was:

```ruby
def latest_timestamp_by_class(cached_class)
  cached_class.maximum(:updated_at).to_i
end
```

Reply:
I've refactored this to a service and added more comments.

Reply:
> Rails's cache key applies to a single record, right? Like cache: @post

that's right. However, if this is per-class and checking again the Rails guides, this seems to then totally duplicate AR's built-in query cache? Are we sure we're not caching what's already cached?

Sources:

* https://guides.rubyonrails.org/caching_with_rails.html#sql-caching
* https://coderwall.com/p/tvybhq/bypassing-activerecord-query-cache

Reply:
Aha, so we can just use the simpler/cleaner method and not bother memoizing it?